datalad/cmd.py

```diff
@@ -141,27 +143,159 @@ def run_gitcommand_on_file_list_chunks(func, cmd, files, *args, **kwargs):
+      asyncio event loop instance. Must support subprocesses on the
+      target platform.
+    cmd : list
+      Command to be executed, passed to `subprocess_exec`.
+    protocol : WitlessProtocol
+      Protocol class to be instantiated for managing communication
+      with the subprocess.
+    stdin : file-like or None
+      Passed to the subprocess as its standard input.
+    kwargs : Pass to `subprocess_exec`, will typically be parameters
+      supported by `subprocess.Popen`.
+
+    Returns
+    -------
+    undefined
+      The nature of the return value is determined by the given
+      protocol class.
+    """
+    lgr.debug('Async run %s', cmd)
+
+    cmd_done = asyncio.Future(loop=loop)
+    factory = functools.partial(protocol, cmd_done)
+    proc = loop.subprocess_exec(
+        factory,
+        *cmd,
+        stdin=stdin,
+        # ask the protocol which streams to capture
+        stdout=asyncio.subprocess.PIPE if protocol.proc_out else None,
+        stderr=asyncio.subprocess.PIPE if protocol.proc_err else None,
+        **kwargs
+    )
+    transport = None
+    try:
+        lgr.debug('Launching process %s', cmd)
+        transport, protocol = await proc
+        lgr.debug('Waiting for process %i to complete', transport.get_pid())
+        await cmd_done
+    finally:
+        # protect against a crash whe launching the process
+        if transport:
+            transport.close()
+
+    return cmd_done.result()
+
+
+class WitlessProtocol(asyncio.SubprocessProtocol):
+    """Subprocess communication protocol base class for `run_async_cmd`
+
+    This class implements basic subprocess output handling. Derived classes
+    like `StdOutCapture` should be used for subprocess communication that need
+    to capture and return output. In particular, the `pipe_data_received()`
+    method can be overwritten to implement "online" processing of process
+    output.
+
+    This class defines a default return value setup that causes
+    `run_async_cmd()` to return a 2-tuple with the subprocess's exit code
+    and a list with bytestrings of all captured output streams.
+    """
+
+    FD_NAMES = ['stdin', 'stdout', 'stderr']
+
+    proc_out = None
+    proc_err = None
+
+    def __init__(self, done_future):
+        # future promise to be fulfilled when process exits
+        self.done = done_future
+        # capture output in bytearrays while the process is running
+        Streams = namedtuple('Streams', ['out', 'err'])
+        self.buffer = Streams(
+            out=bytearray() if self.proc_out else None,
+            err=bytearray() if self.proc_err else None,
+        )
+        self.pid = None
+        super().__init__()
+
+    def connection_made(self, transport):
+        self.transport = transport
+        self.pid = transport.get_pid()
+        lgr.debug('Process %i started', self.pid)
+
+    def pipe_data_received(self, fd, data):
+        if lgr.isEnabledFor(logging.DEBUG):
+            lgr.debug(
```

Comment:
@yarikoptic:  `lgr.debug` seems to be too high level for such type of messages. could you move it under `lgr.log(5, ` or even below?

Reply:
Done @jbpoline